tokio-tcp/src/split.rs

```diff
@@ -0,0 +1,145 @@
+//! `TcpStream` split support.
+//!
+//! A `TcpStream` can be split into a `TcpStreamReadHalf` and a
+//! `TcpStreamWriteHalf` with the `TcpStream::split` method. `TcpStreamReadHalf`
+//! implements `AsyncRead` while `TcpStreamWriteHalf` implements `AsyncWrite`.
+//! The two halves can be used concurrently, even from multiple tasks.
+//!
+//! Compared to the generic split of `AsyncRead + AsyncWrite`, this specialized
+//! split gives read and write halves that are faster and smaller, because they
+//! do not use locks. They also provide access to the underlying `TcpStream`
+//! after split, implementing `AsRef<TcpStream>`. This allows you to call
+//! `TcpStream` methods that takes `&self`, e.g., to get local and peer
+//! addresses, to get and set socket options, and to shutdown the sockets.
+
+use super::TcpStream;
+use bytes::{Buf, BufMut};
+use std::error::Error;
+use std::fmt;
+use std::io;
+use std::net::Shutdown;
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+use tokio_io::{AsyncRead, AsyncWrite};
+
+/// Read half of a `TcpStream`.
+#[derive(Debug)]
+pub struct TcpStreamReadHalf(Arc<TcpStream>);
+
+/// Write half of a `TcpStream`.
+///
+/// Note that in the `AsyncWrite` implemenation of `TcpStreamWriteHalf`,
+/// `poll_shutdown` actually shuts down the TCP stream in the write direction.
+#[derive(Debug)]
+pub struct TcpStreamWriteHalf(Arc<TcpStream>);
+
+pub(crate) fn split(stream: TcpStream) -> (TcpStreamReadHalf, TcpStreamWriteHalf) {
+    let shared = Arc::new(stream);
+    (
+        TcpStreamReadHalf(shared.clone()),
+        TcpStreamWriteHalf(shared),
+    )
+}
+
+/// Error indicating two halves were not from the same stream, and thus could
+/// not be `reunite`d.
+#[derive(Debug)]
+pub struct ReuniteError(pub TcpStreamReadHalf, pub TcpStreamWriteHalf);
+
+impl fmt::Display for ReuniteError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "tried to reunite halves that are not from the same stream"
+        )
+    }
+}
+
+impl Error for ReuniteError {}
+
+impl TcpStreamReadHalf {
+    /// Attempts to put the two "halves" of a `TcpStream` back together and
+    /// recover the original stream. Succeeds only if the two "halves"
+    /// originated from the same call to `TcpStream::split`.
+    pub fn reunite(self, other: TcpStreamWriteHalf) -> Result<TcpStream, ReuniteError> {
+        if Arc::ptr_eq(&self.0, &other.0) {
+            drop(other);
+            Ok(Arc::try_unwrap(self.0).unwrap())
```

Comment:
Could you make this into an `expect` with an explanation for why it should always be ok?

Reply:
Ah, sorry, what I meant here was to explain why you are expecting this unwrap to never fail. I _believe_ it is because you only ever create two instances of the `Arc`, one for the reader and one for the writer, and those `Arc`s are never exposed externally. And so when you drop one here, the other one _must_ be the only remaining one.